### PR TITLE
HOTT-1174 Fixes in suggestions

### PR DESCRIPTION
**Pull request by alexdesi**
### What?
This PR fixes:
1) suggestion called with beginning spaces.
2) 500 Internal Server Error when suggestions are not found.

### Jira link
https://transformuk.atlassian.net/browse/HOTT-1174

### Why?
We don't like bugs!

Screenshot 2021-12-21 at 12 54 10

### CODE REVIEW

File: app/controllers/search_controller.rb [hunk 1]
```diff
@@ -37,7 +37,7 @@ def search
   end
 
   def suggestions
-    search_term = Regexp.escape(params[:term].to_s)
+    search_term = Regexp.escape(params[:term].to_s.strip)
```
Comment (willfish) [inline app/controllers/search_controller.rb:40]: Should we add a test for this to avoid regressions?

File: app/models/search_suggestion.rb [hunk 2]
```diff
@@ -17,8 +17,12 @@ def cached_suggestions_for(term)
     end
 
     def start_with(term)
-      cached_suggestions_for(term).select { |s| s.value =~ /^#{term}/i }
-                                  .first(TradeTariffFrontend.suggestions_count)
+      cached_suggestions = cached_suggestions_for(term)
+
+      return [] unless cached_suggestions
```
Comment (willfish) [inline app/models/search_suggestion.rb]: This isn't needed AFAIK

If the cached_suggestions_for method returns an empty Array when there are none (it definitely should do), then the select will iterate over nothing and return an empty Array and the first method will return an empty array, too.

File: spec/models/search_suggestion_spec.rb [hunk 3]
```diff
@@ -28,5 +28,13 @@
     it 'returns less then 10 values' do
       expect(described_class.start_with('80').length).to be <= 10
     end
+
+    context 'when cached suggestions is nil' do
+      it 'returns empty array' do
+        allow(described_class).to receive(:cached_suggestions_for).and_return(nil)
```
Comment (willfish) [inline spec/models/search_suggestion_spec.rb]: We shouldn't mock the subject under test

File: app/models/search_suggestion.rb (hunk 2 above)
Comment (alexdesi) [reply app/models/search_suggestion.rb]: The issue is that  cached_suggestions can return nil,
in that case we get NoMethodError,
for example when you try to search "+" (the example in the screenshot above)

File: app/controllers/search_controller.rb (hunk 1 above)
Comment (alexdesi) [reply app/controllers/search_controller.rb:40]: yep, I agree

**Review by willfish — APPROVED** [review]

File: app/models/search_suggestion.rb [hunk 4]
```diff
@@ -32,12 +32,13 @@ def cached_suggestions_for_letter(first_letter)
...
 
       grouped_results = all.group_by { |r| r.value.to_s.downcase.slice(0, 1) }
+
       grouped_results.transform_keys! { |letter| "#{cache_prefix}#{letter}" }
       grouped_results["#{cache_prefix}loaded"] = true
 
       Rails.cache.write_multi grouped_results, expires_in: 24.hours
 
-      grouped_results["#{cache_prefix}#{first_letter}"]
+      grouped_results["#{cache_prefix}#{first_letter}"].presence || []
```
Comment (willfish) [inline app/models/search_suggestion.rb:41]: :tada:

**Pull request merged**